cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d_test.cc

```diff
@@ -0,0 +1,162 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cartographer/pose_graph/constraint/cost_function/relative_pose_cost_2d.h"
+
+#include "cartographer/common/make_unique.h"
+#include "cartographer/mapping/internal/optimization/cost_functions/cost_helpers.h"
+#include "gmock/gmock.h"
+#include "google/protobuf/text_format.h"
+
+namespace cartographer {
+namespace pose_graph {
+namespace {
+
+constexpr int kPoseDimension = 3;
+constexpr int kResidualsCount = 3;
+constexpr int kParameterBlocksCount = 2;
+constexpr int kJacobianColDimension = kResidualsCount * kPoseDimension;
+
+using ::google::protobuf::TextFormat;
+using ::testing::ElementsAre;
+using ResidualType = std::array<double, kResidualsCount>;
+using JacobianType = std::array<std::array<double, kJacobianColDimension>,
+                                kParameterBlocksCount>;
+
+class AutoDiffRelativePoseCost {
+ public:
+  explicit AutoDiffRelativePoseCost(
+      const proto::RelativePose2D::Parameters& parameters)
+      : translation_weight_(parameters.translation_weight()),
+        rotation_weight_(parameters.rotation_weight()),
+        first_T_second_(transform::ToRigid2(parameters.first_t_second())) {}
+
+  template <typename T>
+  bool operator()(const T* const start_pose, const T* const end_pose,
+                  T* e) const {
+    const std::array<T, 3> error = mapping::optimization::ScaleError(
+        mapping::optimization::ComputeUnscaledError(first_T_second_, start_pose,
+                                                    end_pose),
+        translation_weight_, rotation_weight_);
+    std::copy(std::begin(error), std::end(error), e);
+    return true;
+  }
+
+ private:
+  const double translation_weight_;
+  const double rotation_weight_;
+  const transform::Rigid2d first_T_second_;
+};
+
+class RelativePoseCost2DTest : public ::testing::Test {
+ public:
+  RelativePoseCost2DTest() {
+    proto::RelativePose2D::Parameters parameters;
+    constexpr char kParameters[] = R"PROTO(
+      first_t_second {
+        translation: { x: 1 y: 1 }
+        rotation: -2.214297
+      }
+      translation_weight: 1
+      rotation_weight: 10
+    )PROTO";
+    EXPECT_TRUE(TextFormat::ParseFromString(kParameters, &parameters));
+
+    auto_diff_cost_ = common::make_unique<RelativePoseCost2D>(parameters);
+    analytical_cost_ = common::make_unique<
+        ceres::AutoDiffCostFunction<AutoDiffRelativePoseCost, kResidualsCount,
+                                    kPoseDimension, kPoseDimension>>(
+        new AutoDiffRelativePoseCost(parameters));
+    for (int i = 0; i < kParameterBlocksCount; ++i) {
+      jacobian_ptrs_[i] = jacobian_[i].data();
+    }
+  }
+
+  std::pair<const ResidualType&, const JacobianType&> EvaluateAnalyticalCost(
```

Comment:
Let's rename this to EvaluateRelativePoseCost2D to make it clear that this is the method we want to test. Please also rename analytical_cost_ accordingly.

Reply:
done